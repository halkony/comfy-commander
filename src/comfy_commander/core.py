"""
Core functionality for Comfy Commander.
"""

import json
from typing import Dict, Any, Optional

import attrs


@attrs.define
class PropertyAccessor:
    """Allows property access and assignment for node properties."""
    
    node: "Node" = attrs.field()
    property_name: str = attrs.field()
    
    def __eq__(self, other: Any) -> bool:
        """Compare property value for equality."""
        return self.node.inputs.get(self.property_name) == other
    
    def __ne__(self, other: Any) -> bool:
        """Compare property value for inequality."""
        return self.node.inputs.get(self.property_name) != other
    
    def __repr__(self) -> str:
        """String representation of the property value."""
        return repr(self.node.inputs.get(self.property_name))
    
    def __setattr__(self, name: str, value: Any) -> None:
        """Handle assignment to the property accessor itself."""
        if name in ["node", "property_name"]:
            super().__setattr__(name, value)
        else:
            # This handles the case where we assign directly to the property accessor
            self.node.inputs[self.property_name] = value
    
    def __call__(self, value: Any = None) -> Any:
        """Allow direct assignment and value retrieval."""
        if value is not None:
            self.node.inputs[self.property_name] = value
            return value
        return self.node.inputs.get(self.property_name)
    
    def set(self, value: Any) -> None:
        """Set the property value."""
        self.node.inputs[self.property_name] = value
    
    @property
    def value(self) -> Any:
        """Get the property value."""
        return self.node.inputs.get(self.property_name)


@attrs.define
class Node:
    """Represents a single node in a ComfyUI workflow."""
    
    id: str = attrs.field()
    data: Dict[str, Any] = attrs.field()
    inputs: Dict[str, Any] = attrs.field(init=False)
    class_type: str = attrs.field(init=False)
    meta: Dict[str, Any] = attrs.field(init=False)
    
    def __attrs_post_init__(self):
        """Initialize derived fields after attrs initialization."""
        self.inputs = self.data.get("inputs", {})
        self.class_type = self.data.get("class_type", "")
        self.meta = self.data.get("_meta", {})
    
    def property(self, name: str) -> PropertyAccessor:
        """Get a property accessor for the node's inputs."""
        return PropertyAccessor(node=self, property_name=name)


@attrs.define
class Workflow:
    """Represents a ComfyUI workflow with nodes and their connections."""
    
    data: Dict[str, Any] = attrs.field()
    nodes: Dict[str, Node] = attrs.field(init=False)
    
    def __attrs_post_init__(self):
        """Initialize nodes after attrs initialization."""
        self.nodes = {}
        for node_id, node_data in self.data.items():
            self.nodes[node_id] = Node(id=node_id, data=node_data)
    
    @classmethod
    def from_file(cls, file_path: str) -> "Workflow":
        """Load a workflow from a JSON file."""
        with open(file_path, 'r') as f:
            workflow_data = json.load(f)
        return cls(data=workflow_data)
    
    def node(self, id: Optional[str] = None, name: Optional[str] = None) -> Node:
        """Get a node by ID or by class_type name."""
        if id is not None:
            if id in self.nodes:
                return self.nodes[id]
            raise KeyError(f"Node with ID '{id}' not found")
        
        if name is not None:
            for node in self.nodes.values():
                if node.class_type == name:
                    return node
            raise KeyError(f"Node with class_type '{name}' not found")
        
        raise ValueError("Either 'id' or 'name' must be provided")
