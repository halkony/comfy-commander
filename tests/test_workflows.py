from comfy_commander import Workflow
import pytest
import os

@pytest.fixture
def api_workflow_file_path():
    return os.path.join(os.path.dirname(__file__), "fixtures", "flux_dev_checkpoint_example_api.json")

class TestWorkflows:
    def test_workflow_node_editable_by_id(self, snapshot, api_workflow_file_path):
        workflow = Workflow.from_file(api_workflow_file_path)
        workflow.node(id="31").property("seed").set(1234567890)
        assert workflow.node(id="31").property("seed").value == 1234567890
        assert workflow == snapshot

    def test_workflow_node_editable_by_name(self, snapshot, api_workflow_file_path):
        workflow = Workflow.from_file(api_workflow_file_path)
        workflow.node(name="KSampler").property("seed").set(1234567890)
        assert workflow.node(name="KSampler").property("seed").value == 1234567890
        assert workflow == snapshot
