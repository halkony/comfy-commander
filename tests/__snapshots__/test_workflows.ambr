# serializer version: 1
# name: TestWorkflows.test_workflow_node_editable_by_id
  Workflow(data={'6': {'inputs': {'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Positive Prompt)'}}, '8': {'inputs': {'samples': ['31', 0], 'vae': ['39', 0]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, '9': {'inputs': {'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, '27': {'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptySD3LatentImage', '_meta': {'title': 'EmptySD3LatentImage'}}, '30': {'inputs': {'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, '31': {'inputs': {'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, '33': {'inputs': {'text': '', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Negative Prompt)'}}, '35': {'inputs': {'guidance': 3.5, 'conditioning': ['6', 0]}, 'class_type': 'FluxGuidance', '_meta': {'title': 'FluxGuidance'}}, '38': {'inputs': {'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, 'class_type': 'DualCLIPLoader', '_meta': {'title': 'DualCLIPLoader'}}, '39': {'inputs': {'vae_name': 'ae.safetensors'}, 'class_type': 'VAELoader', '_meta': {'title': 'Load VAE'}}, '40': {'inputs': {'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}}, nodes={'6': Node(id='6', data={'inputs': {'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Positive Prompt)'}}, inputs={'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, class_type='CLIPTextEncode', meta={'title': 'CLIP Text Encode (Positive Prompt)'}), '8': Node(id='8', data={'inputs': {'samples': ['31', 0], 'vae': ['39', 0]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, inputs={'samples': ['31', 0], 'vae': ['39', 0]}, class_type='VAEDecode', meta={'title': 'VAE Decode'}), '9': Node(id='9', data={'inputs': {'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, inputs={'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, class_type='SaveImage', meta={'title': 'Save Image'}), '27': Node(id='27', data={'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptySD3LatentImage', '_meta': {'title': 'EmptySD3LatentImage'}}, inputs={'width': 1024, 'height': 1024, 'batch_size': 1}, class_type='EmptySD3LatentImage', meta={'title': 'EmptySD3LatentImage'}), '30': Node(id='30', data={'inputs': {'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, inputs={'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, class_type='CheckpointLoaderSimple', meta={'title': 'Load Checkpoint'}), '31': Node(id='31', data={'inputs': {'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, inputs={'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, class_type='KSampler', meta={'title': 'KSampler'}), '33': Node(id='33', data={'inputs': {'text': '', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Negative Prompt)'}}, inputs={'text': '', 'clip': ['40', 1]}, class_type='CLIPTextEncode', meta={'title': 'CLIP Text Encode (Negative Prompt)'}), '35': Node(id='35', data={'inputs': {'guidance': 3.5, 'conditioning': ['6', 0]}, 'class_type': 'FluxGuidance', '_meta': {'title': 'FluxGuidance'}}, inputs={'guidance': 3.5, 'conditioning': ['6', 0]}, class_type='FluxGuidance', meta={'title': 'FluxGuidance'}), '38': Node(id='38', data={'inputs': {'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, 'class_type': 'DualCLIPLoader', '_meta': {'title': 'DualCLIPLoader'}}, inputs={'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, class_type='DualCLIPLoader', meta={'title': 'DualCLIPLoader'}), '39': Node(id='39', data={'inputs': {'vae_name': 'ae.safetensors'}, 'class_type': 'VAELoader', '_meta': {'title': 'Load VAE'}}, inputs={'vae_name': 'ae.safetensors'}, class_type='VAELoader', meta={'title': 'Load VAE'}), '40': Node(id='40', data={'inputs': {'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}, inputs={'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, class_type='LoraLoader', meta={'title': 'Load LoRA'})})
# ---
# name: TestWorkflows.test_workflow_node_editable_by_name
  Workflow(data={'6': {'inputs': {'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Positive Prompt)'}}, '8': {'inputs': {'samples': ['31', 0], 'vae': ['39', 0]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, '9': {'inputs': {'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, '27': {'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptySD3LatentImage', '_meta': {'title': 'EmptySD3LatentImage'}}, '30': {'inputs': {'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, '31': {'inputs': {'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, '33': {'inputs': {'text': '', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Negative Prompt)'}}, '35': {'inputs': {'guidance': 3.5, 'conditioning': ['6', 0]}, 'class_type': 'FluxGuidance', '_meta': {'title': 'FluxGuidance'}}, '38': {'inputs': {'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, 'class_type': 'DualCLIPLoader', '_meta': {'title': 'DualCLIPLoader'}}, '39': {'inputs': {'vae_name': 'ae.safetensors'}, 'class_type': 'VAELoader', '_meta': {'title': 'Load VAE'}}, '40': {'inputs': {'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}}, nodes={'6': Node(id='6', data={'inputs': {'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Positive Prompt)'}}, inputs={'text': 'cute anime girl with massive fluffy fennec ears and a big fluffy tail blonde messy long hair blue eyes wearing a maid outfit with a long black gold leaf pattern dress and a white apron mouth open placing a fancy black forest cake with candles on top of a dinner table of an old dark Victorian mansion lit by candlelight with a bright window to the foggy forest and very expensive stuff everywhere there are paintings on the walls', 'clip': ['40', 1]}, class_type='CLIPTextEncode', meta={'title': 'CLIP Text Encode (Positive Prompt)'}), '8': Node(id='8', data={'inputs': {'samples': ['31', 0], 'vae': ['39', 0]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, inputs={'samples': ['31', 0], 'vae': ['39', 0]}, class_type='VAEDecode', meta={'title': 'VAE Decode'}), '9': Node(id='9', data={'inputs': {'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, inputs={'filename_prefix': 'ComfyUI', 'images': ['8', 0]}, class_type='SaveImage', meta={'title': 'Save Image'}), '27': Node(id='27', data={'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptySD3LatentImage', '_meta': {'title': 'EmptySD3LatentImage'}}, inputs={'width': 1024, 'height': 1024, 'batch_size': 1}, class_type='EmptySD3LatentImage', meta={'title': 'EmptySD3LatentImage'}), '30': Node(id='30', data={'inputs': {'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, inputs={'ckpt_name': 'flux_project0_v40ArtRealismFP8.safetensors'}, class_type='CheckpointLoaderSimple', meta={'title': 'Load Checkpoint'}), '31': Node(id='31', data={'inputs': {'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, inputs={'seed': 1234567890, 'steps': 8, 'cfg': 1, 'sampler_name': 'euler', 'scheduler': 'simple', 'denoise': 1, 'model': ['40', 0], 'positive': ['35', 0], 'negative': ['33', 0], 'latent_image': ['27', 0]}, class_type='KSampler', meta={'title': 'KSampler'}), '33': Node(id='33', data={'inputs': {'text': '', 'clip': ['40', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Negative Prompt)'}}, inputs={'text': '', 'clip': ['40', 1]}, class_type='CLIPTextEncode', meta={'title': 'CLIP Text Encode (Negative Prompt)'}), '35': Node(id='35', data={'inputs': {'guidance': 3.5, 'conditioning': ['6', 0]}, 'class_type': 'FluxGuidance', '_meta': {'title': 'FluxGuidance'}}, inputs={'guidance': 3.5, 'conditioning': ['6', 0]}, class_type='FluxGuidance', meta={'title': 'FluxGuidance'}), '38': Node(id='38', data={'inputs': {'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, 'class_type': 'DualCLIPLoader', '_meta': {'title': 'DualCLIPLoader'}}, inputs={'clip_name1': 't5xxl_fp8_e4m3fn_scaled.safetensors', 'clip_name2': 'clip_l.safetensors', 'type': 'flux', 'device': 'default'}, class_type='DualCLIPLoader', meta={'title': 'DualCLIPLoader'}), '39': Node(id='39', data={'inputs': {'vae_name': 'ae.safetensors'}, 'class_type': 'VAELoader', '_meta': {'title': 'Load VAE'}}, inputs={'vae_name': 'ae.safetensors'}, class_type='VAELoader', meta={'title': 'Load VAE'}), '40': Node(id='40', data={'inputs': {'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}, inputs={'lora_name': 'Hyper-FLUX.1-dev-8steps-lora.safetensors', 'strength_model': 0.1, 'strength_clip': 1, 'model': ['30', 0], 'clip': ['38', 0]}, class_type='LoraLoader', meta={'title': 'Load LoRA'})})
# ---
